### ISSUE TITLE
`reportIncompatibleMethodOverride` false positive on extra overloads in subclass

### ISSUE THREAD

**Issue by Pwuts** [microsoft/pyright#8007]
### Describe the bug
**TL;DR:** false-positive `Override does not handle all overloads of base method`

Pyright reports an incompatible method override when a method on a subclass has more detailed overloads than the parent class. The overloads of the base class are also present on the subclass, but the subclass has extra, more specific overloads. This constitutes a false-positive IMO.

I'm fairly confident that this is a bug, since I can make the error go away by *removing* two overloads on the subclass, which is contradictory.

Interestingly, Pylance `v2023.10.40` (which afaik relies on Pyright) does not report this error (on `strict` mode).

### Code or Screenshots
No errors on this snippet:
```python
class TypeA:
    pass

class TypeB:
    pass

class BaseClass:
    def meth(self, param1: Literal["a", "b"] = "a") -> TypeA | TypeB:
        return TypeA() if param1 == "a" else TypeB()

class SubClass(BaseClass):
    @overload
    def meth(self, param1: Literal["a"]) -> TypeA: ...
    @overload
    def meth(self, param1: Literal["b"]) -> TypeB: ...
    def meth(self, param1: Literal["a", "b"] = "a") -> TypeA | TypeB:
        print(param1)
        return TypeA() if param1 == "a" else TypeB()
```

But the described error appears in this slightly more complicated case:
```python
class TypeB1(TypeB):
    pass

class TypeB2(TypeB):
    pass

class BaseClass2:
    # Group 1
    @overload
    def meth(self, param1: Literal["a"], param2: bool = False) -> TypeA: ...

    # Group 2
    @overload
    def meth(self, param1: Literal["b"], param2: bool = False) -> TypeB: ...

    # Group 3
    @overload
    def meth(self, param1: Literal["a", "b"] = "a", param2: bool = False) -> TypeA | TypeB: ...

    def meth(self, param1: Literal["a", "b"] = "a", param2: bool = False) -> TypeA | TypeB:
        return TypeA() if param1 == "a" else TypeB()

# Overloads grouped in the same way as above, with every group of
# overloads at least equivalent to the same group in the base class:
class SubClass2(BaseClass2):
    # Group 1
    @overload
    def meth(self, param1: Literal["a"], param2: bool = False) -> TypeA: ...
    @overload
    def meth(self, *, param2: bool = False) -> TypeA: ...

    # Group 2
    @overload
    def meth(self, param1: Literal["b"], param2: Literal[False] = False) -> TypeB1: ...
    @overload
    def meth(self, param1: Literal["b"], param2: Literal[True]) -> TypeB2: ...
    @overload
    def meth(self, param1: Literal["b"], param2: bool = False) -> TypeB1 | TypeB2: ...

    # Group 3
    @overload
    def meth(self, param1: Literal["a", "b"] = "a", param2: Literal[False] = False) -> TypeA | TypeB1: ...
    @overload
    def meth(self, param1: Literal["a", "b"], param2: Literal[True]) -> TypeA | TypeB2: ...
    @overload
    def meth(self, param1: Literal["a", "b"] = "a", param2: bool = False) -> TypeA | TypeB1 | TypeB2: ...

    def meth(self, param1: Literal["a", "b"] = "a", param2: bool = False) -> TypeA | TypeB1 | TypeB2:
        print(param1)
        return TypeA() if param1 == "a" else (TypeB2() if param2 else TypeB1())
    # error: Method "meth" overrides class "BaseClass2" in an incompatible manner
    #   Override does not handle all overloads of base method (reportIncompatibleMethodOverride)
```
Removing the extra overloads in groups 1, 2 and 3 of `SubClass2` makes the error disappear.

<details>
<summary><b>A less abstract example</b></summary>

Base class:
```python
class FileStorage(ABC):
    @overload
    @abstractmethod
    def open_file(
        self,
        path: str | Path,
        mode: Literal["r", "w"] = "r",
        binary: Literal[False] = False,
    ) -> TextIO:
        """Returns a readable text file-like object representing the file."""

    @overload
    @abstractmethod
    def open_file(
        self, path: str | Path, mode: Literal["r", "w"], binary: Literal[True]
    ) -> IOBase:
        """Returns a binary file-like object representing the file."""

    @overload
    @abstractmethod
    def open_file(
        self, path: str | Path, *, binary: Literal[True]
    ) -> IOBase:
        """Returns a readable binary file-like object representing the file."""

    @overload
    @abstractmethod
    def open_file(
        self, path: str | Path, mode: Literal["r", "w"] = "r", binary: bool = False
    ) -> TextIO | IOBase:
        """Returns a file-like object representing the file."""
```
Subclass:
```python
class GCSFileStorage(FileStorage):
    @overload
    def open_file(
        self,
        path: str | Path,
        mode: Literal["r", "w"] = "r",
        binary: Literal[False] = False,
    ) -> TextIOWrapper:  # subclass of TextIO
        ...

    # Removing this overload and the next one resolves the error
    @overload
    def open_file(
        self, path: str | Path, mode: Literal["r"], binary: Literal[True]
    ) -> BlobReader:  # subclass of IOBase
        ...

    @overload
    def open_file(
        self, path: str | Path, mode: Literal["w"], binary: Literal[True]
    ) -> BlobWriter:  # subclass of IOBase
        ...

    @overload
    def open_file(
        self, path: str | Path, mode: Literal["r", "w"], binary: Literal[True]
    ) -> BlobWriter | BlobReader:
        ...

    @overload
    def open_file(self, path: str | Path, *, binary: Literal[True]) -> BlobReader:
        ...

    @overload
    def open_file(
        self, path: str | Path, mode: Literal["r", "w"] = "r", binary: bool = False
    ) -> BlobReader | BlobWriter | TextIOWrapper:
        ...

    def open_file(
        self, path: str | Path, mode: Literal["r", "w"] = "r", binary: bool = False
    ) -> BlobReader | BlobWriter | TextIOWrapper:
        """Open a file in the storage."""
        # [implementation omitted]
```
</details>

### Meta
**Pyright version:** v1.1.364 (installed from Pypi)
**Python version:** v3.11.4
**Command line to repro:** `pyright` or `pyright --pythonversion 3.10`

**Comment by erictraut** [comment]
The [Python typing spec](https://typing.readthedocs.io/en/latest/spec/index.html) does not currently provide any guidance about the correct way for a type checker to detect incompatible overrides. This is something I am hoping to address some time within the next several months. Detecting mismatches is especially complex when both the base method and child method are overloaded. The algorithm that pyright currently uses is based on the algorithm used in mypy. 

One thing I'll point out is that the second overload in "group 1" does not really belong in that group. It should logically be part of "group 3" because its signature is a subtype of the signature of the "group 3" method in the base class. It is not a subtype of the "group 1" method in the base class.

I'm going to leave the current algorithm in place until we achieve consensus within the typing community on a standardized approach / algorithm for this case. I'm going to close this issue because I don't have any plans to make a change currently, but I will revisit it in the future.

For now, you can use a `# type: ignore` or `# pyright: ignore` if you have a need for such a complex overload in your code.
